restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java

```diff
@@ -593,4 +597,23 @@ public void setFillInDefaultValues(boolean fillInDefaultValues)
   {
     _fillInDefaultValues = fillInDefaultValues;
   }
+
+  /**
+   * Set a custom {@link MethodAdapterRegistry} in the config.
+   *
+   * @param methodAdapterRegistry a custom to be set in the config.
+   */
+  public void setMethodAdapterRegistry(MethodAdapterRegistry methodAdapterRegistry) {
+    _methodAdapterRegistry = methodAdapterRegistry;
+  }
+
+  /**
+   * @return Return the custom {@link MethodAdapterRegistry} in the config. Return null if no custom
+   * {@link MethodAdapterRegistry} is provided, then the {@link DefaultMethodAdapterRegistry} will be used for setting
+   * up rest.li server.
+   */
+  public MethodAdapterRegistry getMethodAdapterRegistry() {
+    return Optional.ofNullable(_methodAdapterRegistry)
+            .orElse(new DefaultMethodAdapterRegistry(new ErrorResponseBuilder(_errorResponseFormat)));
```

Comment:
I assume we need to default this for backwards compatibility?

Reply:
Not only for backwards compatibility. If users do not specify `MethodAdapterRegistry` in `RestLiConfig`, this guarantees the use of the default one.

Reply:
I'm skeptical about including the default-constructed object in this override getter. I'm not sure if this will cause any real issues, but we're constructing a new `ErrorResponseBuilder` here rather than using the one defined in `BaseRestLiServer`. It may be better to return null (or an `Optional`) here, then construct the default in `BaseRestLiServer` using the existing `ErrorResponseBuilder`.

Reply:
I think there are two folds of you comments: (1) using existing `ErrorResponseBuilder` and (2) having default in override getter (RestLiConfig).

For (1), the reason to construct the default in the config is to make sure it uses the `ErrorResponseFormat` in the config rather than the local one embedded in `ErrorResponseBuilder`. This should be the intention based on the following analysis. So, I think this would be a good opportunity to make this idea explicit.

Only three dependency chains I can find for `BaseRestLiServer`. They are end at `RestLiServer`.
```
BaseRestLiServer <- RestRestLiServer <- RestLiServer
BaseRestLiServer <- RestRestLiServer <- StreamRestLiServer <- AttachmentHandlingRestLiServer <- RestLiServer
BaseRestLiServer <- StreamRestLiServer <- AttachmentHandlingRestLiServer <- RestLiServer
```
In `RestLiServer`, the `ErrorResponseBuilder`, `RestRestLiServer` and `AttachmentHandlingRestLiServer` are initialized as follows.
```
    ErrorResponseBuilder errorResponseBuilder = new ErrorResponseBuilder(config.getErrorResponseFormat());

    _restRestLiServer = new RestRestLiServer(config,
        resourceFactory,
        engine,
        rootResources,
        errorResponseBuilder);
    _streamRestLiServer = new AttachmentHandlingRestLiServer(config,
        resourceFactory,
        engine,
        rootResources,
        errorResponseBuilder);
```
https://github.com/linkedin/rest.li/blob/710acd54c1ce4ec8cc6615ccf1d95250e8a09ed2/restli-server/src/main/java/com/linkedin/restli/server/RestLiServer.java#L112

Given that this particular builder is immutable (builder might not be the best name, but this is a different issue), this suggests that constructing it in the config or using the local one is the same. Therefore, I think the intention is to have the `DefaultMethodAdapterRegistry` using the `ErrorResponseFormat` in the config.


For (2), I could move this out of the config, but in the `BaseRestLiServer` the initialization will not use the local `ErrorResponseBuilder`.
```
_methodBuildersRegistry = Optional.ofNullable(config.getMethodAdapterRegistry())
            .orElse(new DefaultMethodAdapterRegistry(new ErrorResponseBuilder(_errorResponseFormat)));
```

Reply:
You bring up a good point about "making this idea explicit". The issue is that if we don't use the error response builder in the constructor, then the API is misleading. If the constructor takes in an error response builder, then it implies that this is the one that will be used.

In order to preserve consistency, we should go with one of two routes:
1. Remove the `ErrorResponseBuilder` argument from the constructor and the constructor of all subclasses. The config argument is present, so response builders should just be constructed using the error response format config. Since this is backward-compatible (users may define base classes extending BaseRestLiServer and pass up their own response builder). This is backward-incompatible, so instead please deprecate the existing constructors (with javadoc notes) and create new package-private constructors without the error response builder argument.
2. Follow my original suggestion and use the error response builder provided in the constructor.

I suggest going with option 1, since option 2 would result in a config method that returns an `Optional`, and it would be incompatible to go back.

Reply:
Thank you for the suggestions. I updated code with option 1 dce626d